### PR TITLE
allow syringes to set transfer amount

### PR DESCRIPTION
**Pull request by JustinTrotter**
## About the PR 
So I decided to try out playing as the sole chemist with only using syringes. Making the stuff was a breeze and was more engaging than the current pill meta (where you just scatter a massive amount of pills on the table for the public to use all willy-nilly). I was able to give out precise doses and was happy with the results.

The crew, on the other hand, were not happy. They wanted their pills and they wanted them badly, to the point of breaking into my lab making them themselves. I would ask them why they didn't want to bother with the syringes and the major pain point was the speed in administering the drugs. Crew are limited to only 15u per syringe and could only inject 5u per doafter.

Solution I've come up with is to allow syringes to have their solution transfer amount to be configurable just like any other container. It still defaults to 5u, but can be set to 10 or 15u as necessary. This in my opinion buffs syringes just enough to make them more palatable for crew to use.

**Screenshots**

syringe

**Changelog**

:cl:
- add: Syringes can now set how much solution to transfer.

### CODE REVIEW

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs [hunk 1]
```diff
@@ -33,6 +42,47 @@ private void InitializeInjector()
...
+                if ( amount < component.MinimumTransferAmount.Int() || amount > component.MaximumTransferAmount.Int())
+                    continue;
+
+                AlternativeVerb verb = new();
+                verb.Text = Loc.GetString("comp-solution-transfer-verb-amount", ("amount", amount));
+                verb.Category = VerbCategory.SetTransferAmount;
+                verb.Act = () =>
+                {
+                    component.TransferAmount = FixedPoint2.New(amount);
+                    args.User.PopupMessage(Loc.GetString("comp-solution-transfer-set-amount", ("amount", amount)));
```
Comment (ShadowCommander) [inline Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:74]: Maybe make the delay change based on the amount transferred? Something like 1 second per 5u? Still faster than multiple injections, but gives the target a bit more time.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs [hunk 2]
```diff
@@ -33,6 +42,47 @@ private void InitializeInjector()
...
+            if (!EntityManager.TryGetComponent<ActorComponent?>(args.User, out var actor))
+                return;
+
+            // Custom transfer verb
+            AlternativeVerb custom = new();
+            custom.Text = Loc.GetString("comp-solution-transfer-verb-custom-amount");
+            custom.Category = VerbCategory.SetTransferAmount;
+            custom.Act = () => component.UserInterface?.Open(actor.PlayerSession);
+            custom.Priority = 1;
+            args.Verbs.Add(custom);
```
Comment (ShadowCommander) [inline Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:59]: In SS13, it only allows 5, 10, and 15 for syringes. No custom amounts. For smaller amounts, players would use a dropper for 1-5. Should that be replicated here or should it just all be done using syringes?

Comment (JustinTrotter) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:59]: Yeah I'll strip out the custom amount.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs (hunk 1 above)
Comment (JustinTrotter) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:66]: Sounds good. Done.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs [hunk 3]
```diff
@@ -33,6 +42,39 @@ private void InitializeInjector()
         SubscribeLocalEvent<InjectionCancelledEvent>(OnInjectionCancelled);
     }
 
+    private void AddSetTransferVerbs(EntityUid uid, InjectorComponent component, GetVerbsEvent<AlternativeVerb> args)
+        {
+            if (!args.CanAccess || !args.CanInteract || args.Hands == null)
+                return;
```
Comment (0x6273) [inline Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:48]: This function is double-indented

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs [hunk 4]
```diff
@@ -33,6 +42,39 @@ private void InitializeInjector()
...
+                // we want to sort by size, not alphabetically by the verb text.
+                verb.Priority = priority;
+                priority--;
+
+                args.Verbs.Add(verb);
+            }
+        }
+
+
     private static void OnInjectionCancelled(InjectionCancelledEvent ev)
```
Comment (0x6273) [inline Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:78]: Should just be one blank line here.

File: Content.Server/Chemistry/Components/InjectorComponent.cs [hunk 5]
```diff
@@ -64,5 +81,7 @@ public InjectorToggleMode ToggleState
                 Dirty();
             }
         }
+
+        [ViewVariables] public BoundUserInterface? UserInterface => Owner.GetUIOrNull(TransferAmountUiKey.Key);
```
Comment (0x6273) [inline Content.Server/Chemistry/Components/InjectorComponent.cs:85]: Is this needed? Doesn't look like it's used.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs [hunk 6]
```diff
@@ -15,13 +15,22 @@
...
 namespace Content.Server.Chemistry.EntitySystems;
 
 public sealed partial class ChemistrySystem
 {
+
+    /// <summary>
+    ///     Default transfer amounts for the set-transfer verb.
+    /// </summary>
+    public static readonly List<int> DefaultTransferAmounts = new() { 5, 10, 15};
     private void InitializeInjector()
```
Comment (0x6273) [inline Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:31]:
```suggestion
{
    /// <summary>
    ///     Default transfer amounts for the set-transfer verb.
    /// </summary>
    public static readonly List<int> DefaultTransferAmounts = new() { 5, 10, 15};
    
    private void InitializeInjector()
```
Also maybe just call this `TransferAmounts`

Comment (JustinTrotter) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs]: I don't see any difference in the suggested change diff. Updated the name as requested.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs (hunk 3 above)
Comment (JustinTrotter) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs]: Derp.

File: Content.Server/Chemistry/Components/InjectorComponent.cs (hunk 5 above)
Comment (JustinTrotter) [reply Content.Server/Chemistry/Components/InjectorComponent.cs]: RIght. This was used in the custom transfer amount that was stripped out, thanks.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs (hunk 6 above)
Comment (0x6273) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs]: Just whitespace. There should be a blank line before the first method.

**Review by 0x6273 — APPROVED** [review]

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs [hunk 7]
```diff
@@ -33,6 +42,38 @@ private void InitializeInjector()
         SubscribeLocalEvent<InjectionCancelledEvent>(OnInjectionCancelled);
     }
 
+    private void AddSetTransferVerbs(EntityUid uid, InjectorComponent component, GetVerbsEvent<AlternativeVerb> args)
+    {
+        if (!args.CanAccess || !args.CanInteract || args.Hands == null)
+            return;
+
+        if (!EntityManager.TryGetComponent<ActorComponent?>(args.User, out var actor))
```
Comment (0x6273) [inline Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:50]: There's a `TryGetComp` proxy function you could use instead. It does the same it's just shorthand.

**Comment by ShadowCommander** [comment]
Looks good. Just so you know, in the future when you finish part of a review you can hit the Resolve conversation button to minimize it and let the reviewer know that you addressed that part of the review.

File: Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs (hunk 7 above)
Comment (0x6273) [reply Content.Server/Chemistry/EntitySystems/ChemistrySystem.Injector.cs:50]: Tho it looks like `actor` never gets used so it should probably be `HasComp`.

**Pull request merged**